programs/pluto/kernel.c

```diff
@@ -1893,15 +1901,23 @@ static bool setup_half_kernel_state(struct state *st, enum direction direction)
 			DBG_dump_hunk("ESP integrity key:", said_next->integ_key);
 		}
 
-		setup_esp_nic_offload(said_next, c, &nic_offload_fallback);
+		setup_esp_nic_offload(&said_next->nic_offload, c, &nic_offload_fallback);
 
 		bool ret = kernel_ops_add_sa(said_next, replace, st->st_logger);
 
 		if (!ret && nic_offload_fallback &&
-		    said_next->nic_offload_dev != NULL) {
-			/* Fallback to non-nic-offload crypto */
-			said_next->nic_offload_dev = NULL;
-			ret = kernel_ops_add_sa(said_next, replace, st->st_logger);
+		    said_next->nic_offload.dev != NULL) {
+			/* Fallback to nic-offload crypto from packet offload */
+			if (said_next->nic_offload.type == OFFLOAD_PACKET) {
+				said_next->nic_offload.type = OFFLOAD_CRYPTO;
+				ret = kernel_ops_add_sa(said_next, replace, st->st_logger);
+			}
+
+			if (!ret) {
+				/* Fallback to non-nic-offload crypto */
+				said_next->nic_offload.dev = NULL;
+				ret = kernel_ops_add_sa(said_next, replace, st->st_logger);
+			}
```

Comment:
maybe make this conditional on off_auto, and take off_packet and off_yes (off_crypto) to mean only try to do this specific variant (eg to work around unexpected bugs). This way auto could be robust but we could also force to only set crypto or packet offload. This will also give people who want to insist on offload a change to fail the tunnel if they cannot hw offload.

Reply:
the above code is only if off_auto is set (only then nic_offload_fallback is true)